tests/integration/reapply_test.py

```diff
@@ -0,0 +1,140 @@
+#
+# Copyright 2018 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+from subprocess import check_output
+from subprocess import check_call
+
+import gi                           # pylint: disable=import-error
+gi.require_version('NM', '1.0')     # NOQA: F402
+from gi.repository import NM        # pylint: disable=import-error
+from libnmstate import netapplier
+
+from .testlib import statelib
+from .testlib.statelib import INTERFACES
+
+
+_EXTRA_CONFIGS = [
+    {
+        'name': 'ipv4.dad-timeout',
+        'value': '1000'
+    },
+    {
+        'name': 'ipv6.dhcp-hostname',
+        'value': 'libnmstete.example.com'
+    }
+]
+
+
+_INIT_STATE = {
+    "interfaces": [
+        {
+            "name": "eth1",
+            "type": "ethernet",
+            "state": "up",
+            "ipv4": {
+                "address": [
+                    {
+                        "ip": "192.168.122.250",
+                        "prefix-length": 24
+                    }
+                ],
+                "enabled": True
+            },
+            "ipv6": {
+                "address": [
+                    {
+                        "ip": "2001::9",
+                        "prefix-length": 64
+                    }
+                ],
+                "enabled": True
+            },
+            "mtu": 1500
+        },
+    ]
+}
+
+
+
+def _apply_init_config():
+    netapplier.apply(_INIT_STATE)
+
+
+
+def _get_nm_profile_uuid(iface_name):
+    nmcli = NM.Client.new(None)
+    cur_dev = None
+    for dev in nmcli.get_all_devices():
+        if dev.get_iface() == iface_name:
+            cur_dev = dev
+            break
+
+    assert cur_dev
+
+    active_conn = cur_dev.get_active_connection()
+    assert active_conn
+    return active_conn.get_uuid()
+
+
+def test_reapply_preserve_ip_config():
```

Comment:
It have trouble understanding the test, it feels too complex to me.
After you apply the base config (and knowing already the device manipulated), you can just execute `nmcli con modify` directly to add a special extra parameter. No need for a generic dict with these embedded.

In addition, the IPv4 and IPv6 scenarios are different, they should have their own test separately.

You also have libcmd.exec_cmd in you disposal, check the nmstatectl_show_test test module.
You can just check that an extra config value is still there by a simple `a in b` assert.